chakrashim: improve hot path calls

obastemur opened the pull request:
- Context has the cached `null` and `undefined` pointer references. Use them instead of going for type discovery. These two are on the hot path.

 - Number of times this call is successful is merely 0 yet it's in the middle of a `GetObjectData` hot path. If `object` is not proxy, we are able to do `HasProperty` discovery without really entering into script. So.. `JsHasProperty` is reasonably faster than `JsGetProperty + isUndefined`
calls.

MSLaguana commented on deps/chakrashim/src/v8value.cc:
```diff
@@ -45,11 +45,11 @@ static bool IsOfType(const Value* ref, JsValueType type) {
 
 bool Value::IsNull() const {
-  return IsOfType(this, JsValueType::JsNull);
+    return this == jsrt::GetNull();
```
Since we also cache `jsrt::GetTrue()` and `jsrt::GetFalse()` should we also update `Value::IsTrue()` and `Value::IsFalse` to just check for reference equality in the same way? I believe that these would all be equally valid, as long as we continue to have only one script context.

obastemur replied:
It was already being done (long way but done) similarly hence didn't update those. Good idea.

obastemur committed:
chakrashim: improve hot path calls

- Number of times this call is successful is merely 0 yet it's in the
middle of a `GetObjectData` hot path. If `object` is not proxy, we are
able to do `HasProperty` discovery without really entering into script.
So.. `JsHasProperty` is reasonably faster than `JsGetProperty +
isUndefined` calls.

- Use fast pointer compare for True/False/Undefined/Null

jackhorton reviewed (approved):
For completeness, CI link is here: https://ci.nodejs.org/job/chakracore-test/195/

Oz, when you merge this one, could you do the normal node-style merge where you rebase this on top of current master and push it to master as a fast-forward? (rather than clicking the green button on github)
